domain/src/main/java/org/oppia/android/domain/exploration/ExplorationProgressController.kt

```diff
@@ -352,6 +395,114 @@ class ExplorationProgressController @Inject constructor(
+   *  checkpoint with the latest progress and saves it using [ExplorationCheckpointController].
+   *
+   *  This function also waits for the save operation to complete, upon completion this function
+   *  uses the function [processSaveCheckpointResult] to mark the exploration as
+   *  IN_PROGRESS_SAVED or IN_PROGRESS_NOT_SAVED depending upon the result.
+   */
+  private fun saveExplorationCheckpoint() {
+    // Do not save checkpoints if checkpointing is not enabled. This is expected to happen when
+    // the current exploration has been already completed previously.
+    if (!explorationProgress.isLightweightCheckpointingEnabled) return
+    lateinit var checkpointState: CheckpointState
+    val profileId: ProfileId = explorationProgress.currentProfileId
+    val topicId: String = explorationProgress.currentTopicId
+    val storyId: String = explorationProgress.currentStoryId
+    val explorationId: String = explorationProgress.currentExplorationId
+
+    val checkpoint: ExplorationCheckpoint =
+      explorationProgress.stateDeck.createExplorationCheckpoint(
+        explorationProgress.currentExploration.version,
+        explorationProgress.currentExploration.title,
+        oppiaClock.getCurrentTimeMs()
+      )
+
+    val deferred = explorationCheckpointController.recordExplorationCheckpointAsync(
+      profileId,
+      explorationId,
+      checkpoint
+    )
+
+    deferred.invokeOnCompletion {
+      checkpointState = if (it == null) {
+        deferred.getCompleted()
+      } else {
+        oppiaLogger.e("Lightweight checkpointing", "Failed to save checkpoint in exploration", it)
+        // CheckpointState is marked as CHECKPOINT_UNSAVED because the deferred did not
+        // complete successfully.
+        CheckpointState.CHECKPOINT_UNSAVED
+      }
+      processSaveCheckpointResult(
+        profileId,
+        topicId,
+        storyId,
+        explorationId,
+        oppiaClock.getCurrentTimeMs(),
+        checkpointState
+      )
+    }
+  }
+
+  /**
+   * This function process the result obtained upon complete execution of the function
+   * [saveExplorationCheckpoint].
+   *
+   * Marks the exploration as in_progress_saved or in_progress_not_saved if it is not already marked
+   * correctly. This function also updates the checkpoint state of the exploration to the
+   * specified new checkpoint state.
+   *
+   * @param profileId is the profile id currently playing the exploration.
+   * @param topicId is the id of the topic which contains the story with the current exploration.
+   * @param storyId is the id of the story which contains the current exploration.
+   * @param lastPlayedTimestamp timestamp of the time when the checkpoints state for the exploration
+   *     was last updated.
+   * @param newCheckpointState the latest state obtained after saving checkpoint successfully or
+   *     unsuccessfully.
+   */
+  private fun processSaveCheckpointResult(
```

Comment:
Shouldn't this method also notify that the state DataProvider has changed since there's now new data to report to the UI? That part seems important to ensure the UI gets the most up-to-date snapshot of the current state of the controller has possible (to ensure the 'eventual consistency' principle I mentioned in my other comment).

Reply:
As per your earlier comment, we should modify the ephemeral state to store the save state and then use the currentStateDataProvider to pull the save state in the state fragment when the user exits the exploration. But the exploration exits from the ExplorationActivtity and the currentStateDataProvider is in StateFragmentPresenter. So in order to get the save state in ExplorationActivity, we will have to get the currentStateDataProvider in ExplorationActivity, observe it and then pull the save state from there, now since we are reacquiring lock to compute ephemeral state I think we should just check the save state directly without modifying the ephemeral state.

Alternatively, we can update the ephemeral state with the current save state, have a local variable in `stateFragmentPresenter` that updates with the latest save state. And then we use that variable in `ExplorationActivity` by injecting `stateFragmentPresneter` and calling `StateFragmentPresenter.newLocalVariable` but I think this is not the correct way to this.

Reply:
I'm confused--wasn't the entire point of moving the saving to this controller to avoid the complexity of observing the save state in the UI layer? Couldn't we just save a value in StateFragment and call down to it from ExplorationActivity to see if it's set? That seems a lot simpler to me unless I'm missing something.

Reply:
changed the approach to the dataProvider approach.